core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
     }
   }
 
+  public static @Nullable Boolean containsSubstr(@Nullable Object @Nullable [] s0, String s1) {
+    // If s0 has null arguments, it should return TRUE if substring is found, otherwise NULL
+    boolean nullFlag = false;
+    if (s0 == null) {
+      return false;
+    }
+    for (Object obj : s0) {
+      if (obj == null) {
+        nullFlag = true;
+      } else if (obj instanceof Object[]) {
+        return containsSubstr((Object[]) obj, s1);
+      } else if (obj instanceof ArrayList) {
+        return containsSubstr((List) obj, s1);
+      } else if (normalize(obj.toString()).contains(normalize(s1))) {
+        return true;
+      }
+    }
+    return nullFlag ? null : false;
+  }
+
+  public static @Nullable Boolean containsSubstr(List s0, String s1) {
+    // If s0 has null arguments, it should return TRUE if substring is found, otherwise NULL
+    boolean nullFlag = false;
+    for (Object item : s0) {
+      if (item == null) {
+        nullFlag = true;
+      }
+      if (item != null && containsSubstr(item, s1)) {
+        return true;
+      }
+    }
+    return nullFlag ? null : false;
+  }
+
+  public static @Nullable Boolean containsSubstr(String s0, String s1, String s2) {
+    // The third argument specifies the json_scope, either keys, values, or both
+    LinkedHashMap<String, String> map = (LinkedHashMap) JsonFunctions.dejsonize(s0);
+    Set<String> keys = map.keySet();
+    Collection<String> values = map.values();
+    if (s2.equals("JSON_KEYS")) {
+      return keys.contains(s1);
+    } else if (s2.equals("JSON_VALUES")) {
+      return values.contains(s1);
+    } else if (s2.equals("JSON_KEYS_AND_VALUES")) {
+      return keys.contains(s1) || values.contains(s1);
+    } else {
+      throw new IllegalArgumentException("json_scope argument must be one of: \"JSON_KEYS\", "
+          + "\"JSON_VALUES\", \"JSON_KEYS_AND_VALUES\".");
+    }
+  }
+
+  /** SQL {@code CONTAINS_SUBSTR(string, string)} function. */
```

Comment:
Use more descriptive names for arguments. Make the arguments in the javadoc comment (`string`) and Java parameters (`s0`, `s1`) consistent.